File: py-src/fusion/libs/entity/change.py
from __future__ import annotations

from enum import Enum
from typing import Any

from fusion import get_logger
from fusion.libs.entity import Entity, dump_to_dict

log = get_logger(__name__)


class ChangeTypes(Enum):
    EMPTY = 0
    CREATE = 1
    UPDATE = 2
    DELETE = 3


class Change:
    """Represents a single entity-level change using forward/reverse deltas.

    - CREATE: reverse_component is empty, forward_component has full state
    - DELETE: reverse_component has full state, forward_component is empty
    - UPDATE: both have partial diffs (only changed fields)
    """

    __slots__ = ("entity_id", "reverse_component", "forward_component")

    def __init__(
        self,
        entity_id: str,
        reverse_component: dict[str, Any],
        forward_component: dict[str, Any],
    ):
        self.entity_id = entity_id
        self.reverse_component = reverse_component
        self.forward_component = forward_component

    def asdict(self) -> tuple[str, dict[str, Any], dict[str, Any]]:
        """Serialize to the wire format: (entity_id, reverse, forward)."""
        return (self.entity_id, self.reverse_component, self.forward_component)

    def type(self) -> ChangeTypes:
        has_reverse = len(self.reverse_component) > 0
        has_forward = len(self.forward_component) > 0
        if has_reverse and has_forward:
            return ChangeTypes.UPDATE
        elif has_reverse:
            return ChangeTypes.DELETE
        elif has_forward:
            return ChangeTypes.CREATE
        else:
            return ChangeTypes.EMPTY

    @property
    def change_type(self) -> ChangeTypes:
        return self.type()

    @staticmethod
    def create(entity: Entity) -> Change:
        return Change(entity.id, {}, dump_to_dict(entity))

    @staticmethod
    def delete(entity: Entity) -> Change:
        return Change(entity.id, dump_to_dict(entity), {})

    @staticmethod
    def update(old_entity: Entity, new_entity: Entity) -> Change:
        return old_entity.change_from(new_entity)

    def reversed(self) -> Change:
        return Change(self.entity_id, self.forward_component, self.reverse_component)

    def is_create(self) -> bool:
        return self.type() == ChangeTypes.CREATE

    def is_update(self) -> bool:
        return self.type() == ChangeTypes.UPDATE

    def is_delete(self) -> bool:
        return self.type() == ChangeTypes.DELETE

    def is_empty(self) -> bool:
        return self.type() == ChangeTypes.EMPTY

    def __repr__(self) -> str:
        return (
            f"<Change entity_id={self.entity_id} "
            f"type={self.type()} "
            f"reverse_keys={list(self.reverse_component.keys())} "
            f"forward_keys={list(self.forward_component.keys())}>"
        )

File: js-src/src/model/Change.ts
import { Entity, EntityData, SerializedEntityData, dumpToDict } from "./Entity"

export type ChangeComponent = Partial<SerializedEntityData>;

export enum ChangeType {
    EMPTY = 0,
    CREATE = 1,
    UPDATE = 2,
    DELETE = 3
}

export type ChangeData = [
    string, // entity id
    Record<string, any>, // reverse change component
    Record<string, any> // forward change component (apply to get new state)
];

export class Change {
    private _data: ChangeData

    constructor(data: ChangeData) {
        this._data = data
    }

    get data(): ChangeData {
        return this._data
    }

    get entityId(): string {
        return this._data[0]
    }

    get reverseComponent(): ChangeComponent {
        return this._data[1]
    }
    set reverseComponent(value: ChangeComponent) {
        this._data[1] = value
    }

    get forwardComponent(): ChangeComponent {
        return this._data[2]
    }
    set forwardComponent(value: ChangeComponent) {
        this._data[2] = value
    }

    type() {
        const hasReverse = Object.keys(this.reverseComponent).length > 0
        const hasForward = Object.keys(this.forwardComponent).length > 0
        if (hasReverse && hasForward) {
            return ChangeType.UPDATE
        } else if (hasReverse) {
            return ChangeType.DELETE
        } else if (hasForward) {
            return ChangeType.CREATE
        } else {
            return ChangeType.EMPTY
        }
    }

    static create(entity: Entity<EntityData>): Change {
        return new Change([entity.id, {}, dumpToDict(entity)])
    }

    static delete(entity: Entity<EntityData>): Change {
        return new Change([entity.id, dumpToDict(entity), {}])
    }

    static update(oldState: Entity<EntityData>, newState: Entity<EntityData>): Change {
        return oldState.changeFrom(newState)
    }

    reversed(): Change {
        return new Change([this.entityId, this.forwardComponent, this.reverseComponent])
    }

    isCreate(): boolean {
        return this.type() === ChangeType.CREATE
    }

    isDelete(): boolean {
        return this.type() === ChangeType.DELETE
    }

    isUpdate(): boolean {
        return this.type() === ChangeType.UPDATE
    }

    isEmpty(): boolean {
        return this.type() === ChangeType.EMPTY
    }
}
